### PR TITLE
[SPARK-19709][SQL] Read empty file with CSV data source

### PR DESCRIPTION
**Pull request by wojtek-szymanski**
## What changes were proposed in this pull request?

Bugfix for reading empty file with CSV data source. Instead of throwing `NoSuchElementException`, an empty data frame is returned.
 
## How was this patch tested?

Added new unit test in `org.apache.spark.sql.execution.datasources.csv.CSVSuite`

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala [hunk 1]
```diff
@@ -40,7 +41,19 @@ private[csv] object CSVInferSchema {
       csv: Dataset[String],
       caseSensitive: Boolean,
       options: CSVOptions): StructType = {
-    val firstLine: String = CSVUtils.filterCommentAndEmpty(csv, options).first()
+    val lines = CSVUtils.filterCommentAndEmpty(csv, options)
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: Hi @wojtek-szymanski I think we should not rely on exception handling. I can think of `take(1).headOption` but we could use shorten one if you know any other good way. What do you think about this?

Comment (wojtek-szymanski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVInferSchema.scala]: You are absolutely right. Relying on exception handling is smelly, while `Option` gives more opportunities. I also see no difference from performance point of view, since both `first()` and `take(1)` call the the same function `head(1)`.

**Comment by wojtek-szymanski** [comment]
@HyukjinKwon can somebody allow to test this PR as there are no more comments?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 2]
```diff
@@ -133,8 +133,19 @@ object TextInputCSVDataSource extends CSVDataSource {
...
       inputPaths: Seq[FileStatus],
       parsedOptions: CSVOptions): Option[StructType] = {
-    val csv: Dataset[String] = createBaseDataset(sparkSession, inputPaths, parsedOptions)
-    val firstLine: String = CSVUtils.filterCommentAndEmpty(csv, parsedOptions).first()
+    val csv = createBaseDataset(sparkSession, inputPaths, parsedOptions)
+    CSVUtils.filterCommentAndEmpty(csv, parsedOptions)
+      .take(1)
+      .headOption
+      .map(firstLine => infer(sparkSession, parsedOptions, csv, firstLine))
+      .orElse(Some(StructType(Seq())))
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: Could we maybe just match it to [CSVDataSource.scala#L204-L224](https://github.com/wojtek-szymanski/spark/blob/bdf189087c52e8934cee4ee5563dffea9dde6a99/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala#L204-L224) just for consistency for now? 

Personally, I thought such chaining makes hard to read the codes sometimes. Maybe, we could consider the code de-duplication about this in another PR. If would be easier if they looks similar at least.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala [hunk 3]
```diff
@@ -305,13 +305,21 @@ class CSVSuite extends QueryTest with SharedSQLContext with SQLTestUtils {
...
-      .schema(StructType(List(StructField("column", StringType, false))))
+      .schema(StructType(List(StructField("column", StringType, nullable = false))))
       .load(testFile(emptyFile))
 
-    assert(result.collect.size === 0)
+    assert(result.collect().isEmpty)
     assert(result.schema.fieldNames.size === 1)
   }
 
+  test("test with empty file without schema") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Let's re-use the test in [CSVSuite.scala#L1083](https://github.com/wojtek-szymanski/spark/blob/bdf189087c52e8934cee4ee5563dffea9dde6a99/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala#L1083).

We could..

```scala
  test("Empty file produces empty dataframe with empty schema") {
    Seq(false, true).foreach { wholeFile =>
      val df = spark.read.format("csv")
        .option("header", true)
        .option("wholeFile", true)
        .load(testFile(emptyFile))

        assert(df.schema === spark.emptyDataFrame.schema)
        checkAnswer(df, spark.emptyDataFrame)
      }
    }
  }
```

**Comment by HyukjinKwon** [comment]
Hi @wojtek-szymanski, these are all from me. Let me cc @cloud-fan as my PRs related with this were reviewed by him and I guess I can't trigger the test.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala (hunk 3 above)
Comment (wojtek-szymanski) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala]: Good idea, done

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala (hunk 2 above)
Comment (wojtek-szymanski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: I would suggest that we use pattern matching in order to make it more verbose and avoid code like this:

        if (maybeFirstRow.isDefined) {
            val firstRow = maybeFirstRow.get

I also touched `WholeFileCSVDataSource` to unify both implementations. What's your opinion?
Regarding code de-duplication, I fully agree, that it should be done in separate PR.

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 4]
```diff
@@ -133,20 +133,24 @@ object TextInputCSVDataSource extends CSVDataSource {
...
-    val caseSensitive = sparkSession.sessionState.conf.caseSensitiveAnalysis
-    val header = makeSafeHeader(firstRow, caseSensitive, parsedOptions)
-    val tokenRDD = csv.rdd.mapPartitions { iter =>
-      val filteredLines = CSVUtils.filterCommentAndEmpty(iter, parsedOptions)
-      val linesWithoutHeader =
-        CSVUtils.filterHeaderLine(filteredLines, firstLine, parsedOptions)
-      val parser = new CsvParser(parsedOptions.asParserSettings)
-      linesWithoutHeader.map(parser.parseLine)
+    val csv = createBaseDataset(sparkSession, inputPaths, parsedOptions)
+    CSVUtils.filterCommentAndEmpty(csv, parsedOptions).take(1).headOption match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: why call `take(1)`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 5]
```diff
@@ -190,28 +194,26 @@ object WholeFileCSVDataSource extends CSVDataSource {
...
-        UnivocityParser.tokenizeStream(
-          CodecStreams.createInputStreamWithCloseResource(lines.getConfiguration, lines.getPath()),
-          parsedOptions.headerFlag,
-          new CsvParser(parsedOptions.asParserSettings))
-      }
-      Some(CSVInferSchema.infer(tokenRDD, header, parsedOptions))
-    } else {
-      // If the first row could not be read, just return the empty schema.
-      Some(StructType(Nil))
+    }.take(1).headOption match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: same question

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala [hunk 6]
```diff
@@ -190,28 +194,28 @@ object WholeFileCSVDataSource extends CSVDataSource {
...
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: IMHO, `Option.isDefine` with `Option.get`, `Option.map` with `Option.getOrElse` and `Option` with  `match case Some... case None` all might be fine. But, how about minimising the change by matching the above one to `Option.isDefine` with `Option.get`? Then, it would not require the changes here.

Comment (wojtek-szymanski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: I would leave it as it is, since pattern matching still looks a bit clearer than conditionals. If minimizing changes is so critical, I can revert the previous version here and replace pattern matching with conditionals in my fix , @cloud-fan please advise.

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: I don't have a strong preference, this looks fine

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: All three patterns I mentioned are being used across the code base. There is no style guide for this both in https://github.com/databricks/scala-style-guide and http://spark.apache.org/contributing.html

In this case, matching new one to other similar ones is a better choice to reduce changed lines, rather than doing the opposite. Personal taste might be secondary.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: https://github.com/apache/spark/pull/17068#discussion_r104356866 did not show up when I write my comment. I am fine as is. I am not supposed to decide this.

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala (hunk 6 above)
Comment (wojtek-szymanski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: @HyukjinKwon @cloud-fan many thanks for your effort. I really appreciate it and I will take it into account when working with the codebase.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVDataSource.scala]: Thank you both for bearing with me.
